source/pkcs11.c

```diff
@@ -12,3 +20,359 @@
+}
+
+/* Callback for PKCS#11 library to unlock a mutex.
+ * Described in PKCS11-base-v2.40 section 3.7 */
+static CK_RV s_pkcs11_unlock_mutex(CK_VOID_PTR mutex_ptr) {
+    if (mutex_ptr == NULL) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    struct aws_mutex *mutex = mutex_ptr;
+    if (aws_mutex_unlock(mutex)) {
+        AWS_LOGF_ERROR(AWS_LS_IO_PKCS11, "PKCS#11 LockMutex() failed, error %s", aws_error_name(aws_last_error()));
+
+        /* NOTE: Cryptoki has a CKR_MUTEX_NOT_LOCKED error code.
+         * But posix doesn't treat this as an error and neither does windows so ¯\_(ツ)_/¯
+         * If aws_mutex_unlock() failed here, it was something else. */
+        return CKR_GENERAL_ERROR;
+    }
+
+    return CKR_OK;
+}
+
+struct aws_pkcs11_lib {
+    struct aws_ref_count ref_count;
+    struct aws_allocator *allocator;
+
+    struct aws_shared_library shared_lib;
+
+    CK_FUNCTION_LIST_PTR function_list;
+
+    /* If true, C_Finalize() should be called when last ref-count is released */
+    bool should_finalize;
+};
+
+/* Invoked when last ref-count is released. Free all resources.
+ * Note that this is also called if initialization fails half-way through */
+static void s_pkcs11_lib_destroy(void *user_data) {
+    struct aws_pkcs11_lib *pkcs11_lib = user_data;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_PKCS11,
+        "id=%p: Unloading PKCS#11. C_Finalize:%s",
+        (void *)pkcs11_lib,
+        pkcs11_lib->should_finalize ? "yes" : "omit");
+
+    if (pkcs11_lib->should_finalize) {
+        CK_RV rv = pkcs11_lib->function_list->C_Finalize(NULL);
+        if (rv != CKR_OK) {
+            /* Log about it, but continue cleaning up */
+            s_raise_ck_error(pkcs11_lib, "C_Finalize", rv);
+        }
+    }
+
+    aws_shared_library_clean_up(&pkcs11_lib->shared_lib);
+    aws_mem_release(pkcs11_lib->allocator, pkcs11_lib);
+}
+
+struct aws_pkcs11_lib *aws_pkcs11_lib_new(
+    struct aws_allocator *allocator,
+    const struct aws_pkcs11_lib_options *options) {
+
+    /* Create the struct */
+    struct aws_pkcs11_lib *pkcs11_lib = aws_mem_calloc(allocator, 1, sizeof(struct aws_pkcs11_lib));
+    aws_ref_count_init(&pkcs11_lib->ref_count, pkcs11_lib, s_pkcs11_lib_destroy);
+    pkcs11_lib->allocator = allocator;
+
+    /* Load the library. */
+
+    /* need a null-terminated string to call next function,
+     * or NULL if going to search the current application for PKCS#11 symbols. */
+    struct aws_string *filename_storage = NULL;
+    const char *filename = NULL;
+    if (options->filename.len > 0) {
+        filename_storage = aws_string_new_from_cursor(allocator, &options->filename);
+        filename = aws_string_c_str(filename_storage);
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_PKCS11,
+        "Loading PKCS#11. file:'%s' C_Initialize:%s",
+        filename ? filename : "<MAIN PROGRAM>",
+        options->omit_initialize ? "omit" : "yes");
+
+    if (aws_shared_library_init(&pkcs11_lib->shared_lib, filename)) {
```

Comment:
Is it ok to pass NULL filename here?

Reply:
NULL is allowed.
You do this if the symbol can be found in THIS PROGRAM
So like, if you link against libsofthsm2.so file at build time, so the OS automatically loads it along with your program, then you can do: `dlsym(dlopen(NULL), "C_GetFunctionList")` to get at the symbols at runtime.

(I had to do experiments to confirm it worked this way)

Reply:
It's ok to load the library multiple times. from the [docs](https://man7.org/linux/man-pages/man3/dlopen.3.html)
> The dynamic linker maintains reference counts for object handles, so a dynamically loaded shared object is not deallocated until dlclose() has been called on it as many times as dlopen() has succeeded on it